File: go/nth-prime/nth_prime.go
// Package prime
package prime

import "errors"

var primes = []int{2}

func isPrime(i int) bool {
	for _, n := range primes {
		if i % n == 0 {
			return false
		}
	}
	return true
}

func Nth(n int) (int, error) {
	if n < 1 {
		return 0, errors.New("n must be 1 or larger")
	}

	for i := 3; len(primes) < n; i += 2 {
		if isPrime(i) {
			primes = append(primes, i)
		}
	}
	return primes[n - 1], nil
}

File: python/nth-prime/nth_prime.py
"""Compute primes."""

import math
from typing import Generator

def primes() -> Generator[int, None, None]:
    """Prime number generator."""
    # Yield the only even prime then only check odd values.
    yield 2
    found: list[int] = []
    cur = 1
    while True:
        # Test odd values: 3, 5, ...
        cur += 2
        is_prime = True
        # Limit prime checks to sqrt(value)
        limit = math.sqrt(cur)
        for factor in found:
            # If there are no factors <= sqrt(val) then val is prime.
            if factor > limit:
                break
            if cur % factor == 0:
                is_prime = False
                break
        if is_prime:
            found.append(cur)
            yield cur


def prime(number: int) -> int:
    """Return the n'th prime."""
    if number <= 0:
        raise ValueError("there_is_no_zeroth_prime")
    gen = primes()
    return [next(gen) for _ in range(number)][-1]
